### PR TITLE
DAOS-2281 dfuse: Add setattr() callback.

### PR DESCRIPTION
**Pull request by ashleypittman**

### CODE REVIEW

File: src/include/daos_fs.h [hunk 1]
```diff
@@ -443,6 +443,26 @@ dfs_stat(dfs_t *dfs, dfs_obj_t *parent, const char *name,
...
+ * \param[in]	obj	Open object (File, dir or syml) to modify.
+ * \param[in,out]
+ *		stbuf	[in]: Stat struct with the members set.
+ *			[out]: Stat struct with all valid members filled.
+ * \param[in]	flags	Bitmask of flags to set
+ *
+ * \return		0 on Success. Negative errno on Failure.
+ */
+int
+dfs_osetstat(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags);
```
Comment (mchaarawi) [inline src/include/daos_fs.h]: I'd rather have this be just a set rather a set and get (see comments in the function).
Does dfuse need dfs_osetattr()? or can this be a setattr with a parent and a name?

File: src/include/daos_fs.h [hunk 2]
```diff
@@ -443,6 +443,26 @@ dfs_stat(dfs_t *dfs, dfs_obj_t *parent, const char *name,
...
+ * are filled.
+ *
+ * \param[in]	dfs	Pointer to the mounted file system.
+ * \param[in]	obj	Open object (File, dir or syml) to modify.
+ * \param[in,out]
+ *		stbuf	[in]: Stat struct with the members set.
+ *			[out]: Stat struct with all valid members filled.
+ * \param[in]	flags	Bitmask of flags to set
+ *
+ * \return		0 on Success. Negative errno on Failure.
```
Comment (mchaarawi) [inline src/include/daos_fs.h]: not negative errno anymore on failure

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -2350,6 +2350,100 @@ dfs_chmod(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode)
...
+		iods[i].iod_type	= DAOS_IOD_SINGLE;
+	}
+
+	rc = daos_obj_update(oh, th, &dkey, akeys_nr, iods, sgls, NULL);
+	if (rc) {
+		D_ERROR("Failed to update attr (rc = %d)\n", rc);
+		D_GOTO(out, rc = -daos_der2errno(rc));
+	}
+
+	rc = entry_stat(dfs, th, oh, obj->name, stbuf);
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c]: Ideally this would all be read in the same transaction, otherwise there's a potential race condition with rename that we'd need to check for and decide how to handle.  I did initially implement it as needing a separate ostat() call and can move it back to that.

File: src/include/daos_fs.h (hunk 1 above)
Comment (ashleypittman) [reply src/include/daos_fs.h]: I'm fine with changing it to a set only although that removes our ability to handle a race condition.  This does need to be dfs_osetattr() as if it was a setattr we'd need to pass in a oid to verify as part of the operation which brings up back to the same place as osetattr(), but with a worse API.

File: src/include/daos_fs.h (hunk 2 above)
Comment (ashleypittman) [reply src/include/daos_fs.h]: Will fix.

File: src/client/dfs/dfs.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c]: right, if we want to do the revalidate before the access, we would need to read the dkey (probably just the oid, not the whole entry), and verify that with the oid from the object. But we can read the whole dkey and in that case we can return the stat, however for the weaker consistency mode, we don't want to revalidate since we won't use transactions there (in that case you can validate the oid, rename happens, then you do setattr, and if you don't use tx, it won't be detected).
This should apply to other APIs like ostat itself, all the xattr, etc.

So i guess what im saying, we don't want to impose through the API this revalidate step. But if you think in most use cases, a stat is required after a setattr, we can make it in/out with an optional flag to stat or not. in that case i would still call the function dfs_osetattr.

[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c]: Look at the rest of the PR, there's only one caller to d_osetstat() and it needs the attr afterwards.

This interface itself doesn't enforce use of revalidate, but gives a performance improvement if it is being done as more data would be read but it would be the same number of round-trips.

Calling ostat() afterwards where the ostat itself did a revalidate would also work, and would allow us to detect the race condition, but would not allow us to protect against it.

[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c]: OK got it, let's keep the API as is, and we could investigate the revalidate when transactions are done.
the name of the function is still weird in my opinion, and setattr() would be better ;-)

[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c]: Will change, I don't like the current name either.

**Comment by ashleypittman** [comment]
I did think about this a bit more last night, and there would be a good case for modifying the current dfs_chmod() function to be a wrapper around this function, where it would do a lookup,osetattr,close - and in that case we would not need osetattr to perform a fetch of the attributes (the high level fuse interface just calls getattr immediately after the setattr functions).

Looking at the code this also doesn't handle truncate although it worked for the use-cases I tested.  I'll try and trigger this code path today but probably attempt that in a separate PR.

**Comment by ashleypittman** [comment]
> Looking at the code this also doesn't handle truncate although it worked for the use-cases I tested. I'll try and trigger this code path today but probably attempt that in a separate PR.

This code is doing what I'd expect on truncate, which is logging it as a unsupported flag, and returning "Operation not supported"

File: src/client/dfs/dfs.c [hunk 4]
```diff
@@ -2431,14 +2445,16 @@ dfs_osetattr(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
 	rc = daos_obj_update(oh, th, &dkey, akeys_nr, iods, sgls, NULL);
 	if (rc) {
 		D_ERROR("Failed to update attr (rc = %d)\n", rc);
-		D_GOTO(out, rc = -daos_der2errno(rc));
+		D_GOTO(err, rc = daos_der2errno(rc));
 	}
 
+out:
+
 	rc = entry_stat(dfs, th, oh, obj->name, stbuf);
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c]: we can optimize this if the user is setting the size on a file, since we don't need to get the file size in the stat anymore. set/get size is a little expensive and it seems we can avoid a get size in the stat if we have set the size already. it's fine to leave it to a different patch though.

File: src/client/dfs/dfs.c [hunk 5]
```diff
@@ -2431,14 +2445,16 @@ dfs_osetattr(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
...
 	if (rc)
-		D_GOTO(out, rc);
+		D_GOTO(err, rc);
 
-out:
+err:
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c]: probably better to label those out_stat and out_obj, since it's not always err case here

File: src/client/dfs/dfs.c [hunk 6]
```diff
@@ -2431,14 +2445,16 @@ dfs_osetattr(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
...
-		D_GOTO(out, rc = -daos_der2errno(rc));
+		D_GOTO(err, rc = daos_der2errno(rc));
 	}
 
+out:
+
 	rc = entry_stat(dfs, th, oh, obj->name, stbuf);
 	if (rc)
-		D_GOTO(out, rc);
+		D_GOTO(err, rc);
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c]: don't need a check/goto really..

File: src/client/dfs/dfs.c [hunk 7]
```diff
@@ -2408,13 +2409,26 @@ dfs_osetattr(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
 		flags &= ~DFS_SET_ATTR_MTIME;
 		i++;
 	}
+	if (flags & DFS_SET_ATTR_SIZE) {
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c]: Done

**Comment by jolivier23** [comment]
I believe the csum intermittent failure in evtree tests has been fixed in latest master.  Can you verify this is on latest master?

**Comment by ashleypittman** [comment]
> I believe the csum intermittent failure in evtree tests has been fixed in latest master. Can you verify this is on latest master?

It was on latest master this morning, but I've done a merge and picked up two additional commits.
